### ISSUE TITLE
version 2.2.0 makes groups add hang up

### ISSUE THREAD

**Issue by gilesbradshaw** [SignalR/SignalR#3387]
Since upgrading to 2.2.0 when my hubs do groups add it hangs up till I get a task cancelled error

**Comment by davidfowl** [comment]
Is this when you're scaled out or locally?

**Comment by gilesbradshaw** [comment]
I am scaled out using Redis

**Comment by DamianEdwards** [comment]
@gilesbradshaw this is exactly one of the bugs we were trying to fix in 2.2.0, in that previously this *never* worked when doing groups add in scaleout, so this is certainly surprising. Do you have a reduced reproduction project/code you can share with us?

**Comment by sjjohnson** [comment]
To repro, call Groups.Add for more than 5 groups on the same connection.  The 6th and subsequent calls will hang and the task will be cancelled.  Additionally, the first 5 groups you connect to will be disconnected the 6th time you call Groups.Add.

**Comment by sjjohnson** [comment]
By the way, I am running a very simple setup.  Local.  One hub.  2 clients.  One client needs to monitor dozens of groups, so calls Groups.Add dozens of times. The other client just monitors one group, so only calls Groups.Add once.  The 2nd client works fine, even when the first client loses the connection to the first 5 groups it successfully subscribed to.

**Comment by sjjohnson** [comment]
My problem appears to only happen when I'm calling from a JS client. My C# client can add the connection to many groups. I tried 10 and stopped there.

**Comment by gilesbradshaw** [comment]
That's interesting and could be the same as I experienced - apologies for no repro yet just so busy

**Comment by sjjohnson** [comment]
To narrow it down farther, my case only repro's when I use {jsonp: true} on connection.start in the JS client.  If I'm not using JSONP, the JS client can add more connections (I did not find an upper limit).

**Comment by Quethrosar** [comment]
I can confirm this is an issue, but mine happens randomly with just 1 group.
I am using redis for scaleout.

**Comment by johnrutherford** [comment]
I'm having a similar issue with 2.2.0. I'm using Scaleout with RabbitMQ and only adding the connection to 2 groups.

It seems to start happening after a few different clients connect. Most of the time adding the connection to the groups takes 1-2ms, but after a while, it starts taking over 30 seconds and the task is cancelled.

```c#
public override async Task OnConnected()
{
    Stopwatch sw = new Stopwatch();
    sw.Start();

    try
    {
        Task task1 = Groups.Add(Context.ConnectionId, GetCompanyGroupName(Context.User));

        Task task2 = Groups.Add(Context.ConnectionId, GetUserGroupName(Context.User));

        await Task.WhenAll(task1, task2);

        sw.Stop();
    }
    catch (Exception ex)
    {
        sw.Stop();

        string message = String.Format("Exception in OnConnected. " +
            "OnConnected time elapsed: {0}ms", sw.ElapsedMilliseconds);
        Log.LogError(message, ex);
    }

    Log.LogTrace("OnConnected time elapsed: {0}ms", sw.ElapsedMilliseconds);
}
```

**Comment by DamianEdwards** [comment]
Do you still see the issue if you simply await each `Groups.Add` call? It's odd to attempt the adds in parallel (which is what you've done here). Also note that `Stopwatch` isn't accurate across async/await due to it relying on individual CPU core timers. For your case I'd just use `DateTime` deltas:

``` C#
public override async Task OnConnected()
{
    var started = DateTime.UtcNow;

    try
    {
        await Groups.Add(Context.ConnectionId, GetCompanyGroupName(Context.User));
        await Groups.Add(Context.ConnectionId, GetUserGroupName(Context.User));
    }
    catch (Exception ex)
    {
        var message = string.Format("Exception in OnConnected. " +
            "OnConnected time elapsed: {0}ms", (DateTime.UtcNow - started).TotalMilliseconds);
        Log.LogError(message, ex);
    }

    Log.LogTrace("OnConnected time elapsed: {0}ms", (DateTime.UtcNow - started).TotalMilliseconds);
}
```

**Comment by johnrutherford** [comment]
@DamianEdwards, I'll try that and follow up with the results. I realized I was doing the adds in parallel, but yeah, there probably isn't any benefit to doing that.

Good to know about Stopwatch and async/await.

Thanks!

**Comment by johnrutherford** [comment]
It looks like awaiting each add separately fixed the issue for me. Thanks!

**Comment by Jehele** [comment]
@DamianEdwards: about using more Groups.Add calls in paralel as johnrutherford wrote (see copied code below): could you please explain why it's odd? My view is that when I await every Add one by one,  each await means to send request to a (Redis or Sql) server and to wait for a response =  summary time is longer than when I send all requests "together" and wait for all answers. Or do I miss something?
 
```
Task task1 = Groups.Add(Context.ConnectionId, GetCompanyGroupName(Context.User));
Task task2 = Groups.Add(Context.ConnectionId, GetUserGroupName(Context.User));
await Task.WhenAll(task1, task2);
```

**Comment by DamianEdwards** [comment]
@Jehele it depends on the implementation of the scale-out provider. If it doesn't ensure message ordering from the point of calling `Send` then you can easily get into a situation where your parallel group adds overlap each other (the "second" overtakes the "first"). This then results in the client subscriber receiving the later add command first, meaning it will never process the first cmd, hence the ack times out.

The Redis and SQL providers are written to prevent this in that the publish operations are atomic within the backplane server itself (increment of sequence ID and delivery of message).

**Comment by Xiaohongt** [comment]
@sjjohnson, in your repro, when the first client loses the connection in that case, could you check if raised event OnDisconnected paramter stopCalled is just false for every time repro?

**Comment by Xiaohongt** [comment]
@sjjohnson, in your repro, when the first client loses the connection in that case, could you check if raised event OnDisconnected paramter stopCalled is just false for every time repro?

**Comment by Xiaohongt** [comment]
So far I can repro on longPolling without Redis scale-out, when calling Groups.Add() many times with await or wait(), then the tasks could be canceled, and the connection become dead, at this point KA raise Disconnect event with stopCalled false, and increase ConnectionsDisconnected perf counter, but the same connection will be added back again and the connection will continue to work.

**Comment by buybackoff** [comment]
Task cancelled issue on `await Groups.Add(groupId, Context.ConnectionId);` still exists with WebSockets on Win8.1, without scale out, inside IIS express.

**Comment by IanYates** [comment]
@buybackoff - I seem to have a similar issue on occasion.  I'm not doing any scale out, running on IIS Express Win 8.1 using VS 2015 and latest SignalR 2.x from NuGet.  It seems to happen a lot when I have many groups being added.  I'm awaiting each Groups.Add call.  It's inconsistent and so far has been difficult to diagnose as it's difficult to reproduce (using my app anyway).  I'll report back if I get further.

**Comment by niemyjski** [comment]
@IanYates I'm also seeing this with our app ( https://github.com/SignalR/SignalR/issues/3584 ). I've tried everything. If I don't await the Group.Add, I never get the ack (I think) as it doesn't timeout, but I never get any frames.

**Comment by IanYates** [comment]
One thing I've done, and it seems to help, is avoid the advice that seems to be out there that you can register your hubs in your IoC containers as singletons.  Reading through the SignalR code that's just not the case as the Groups property, along with a couple of others, are set via property setters on the Hub instance after its created.

So now I register them in my IoC container (Castle Windsor) as transients.  My SignalR-Windsor bridging code checks to see if it's a Hub instance that's being resolved, and if so, it releases it from Windsor immediately.  That's of no consequence since the Dispose() call really does nothing but it means Windsor isn't holding on to the reference and the property injection that SignalR does is safe.

My suspicion is that, under load, the context of one connection using the Hub class instance was being used by code servicing another connection concurrently.  Yuk :)

The next SignalR is apparently going to have better support for SignalR releasing hub instances when it's done with them as per #2908

**Comment by niemyjski** [comment]
@IanYates Yeah, I made that change now and we just use the connection manager. But still seeing issues.

**Comment by brunobertechini** [comment]
@IanYates  'checks to see if it's a Hub instance that's being resolved, and if so, it releases it from Windsor immediately'

Do you mean: if the serviceType being resolved is a Hub, releases it from Windsor and let the container resolver a new one ? (always transient) ?

Bruno

**Comment by IanYates** [comment]
Yep.  So I let Windsor create it but then immediately tell Windsor to release it - all hubs are transient.  

That gets my dependencies injected (all of which are singletons in Windsor so none of them would themselves be released) but Windsor isn't holding on to the Hub and SignalR is free to just let the garbage collector clean things up.


It's a bit of "here be dragons" area of my code - simple but unexpected.  But making my Hubs transient, and thus getting new instances each time, allows the SignalR property setter code to work without multiple SignalR threads stomping on each other if they shared a single instance of the Hub.  This could be called out much more clearly in the documentation.  Dependency injection via properties, imho, is always a bit icky but I can see why it was done.  David Fowler said ages ago they were hoping to rework the IoC integration with SignalR 3 so I'm happy with my workaround until that arrives later this year.

**Comment by moozzyk** [comment]
In SignalR 3 the SignalR Server will use the built-in DI infrastructure used by ASP.NET Core.

**Comment by johnrutherford** [comment]
This is still happening pretty regularly for us. We are not using dependency injection for Hub instances as others have mentioned. 

As a note, it would be nice to have a fix for this with the 2.x release. It's not going to be easy to switch everything over to ASP.NET Core if this is fixed only in SignalR 3.

**Comment by patrickcarnahan** [comment]
We also see similar behavior on occasion -- hang in Groups.Add(..) for approximately 30 seconds and then an OperationCanceledException bubbles out. I have seen this once in my local environment with almost no load, if it occurs again I will capture the traces and see if it helps root cause the issue.

**Comment by patrickcarnahan** [comment]
I have managed to capture a service profiler trace of this issue on one of our productions servers. Best I can tell it's timing out receiving the ACKs, but this was not previously an issue and we always await our calls to Groups.Add(..). I understand why messages coming out of order would be problematic for SignalR streams, but not sure why that would start happening or happen sporadically on a service bus backplane.

@DamianEdwards I noticed you specifically called out SQL and Redis backplanes are being immune to this issue but said nothing of the service bus provider. Was that just an oversight?

**Comment by aspnet-hello** [comment]
This issue has been closed as part of issue clean-up as described in https://blogs.msdn.microsoft.com/webdev/2018/09/17/the-future-of-asp-net-signalr/. If you're still encountering this problem, please feel free to re-open and comment to let us know! We're still interested in hearing from you, the backlog just got a little big and we had to do a bulk clean up to get back on top of things. Thanks for your continued feedback!

**Comment by byguid** [comment]
Still experiencing the same issue on 2.4.1 with Redis scale out. If I disable backplane - works just fine.
Was this issue addressed in subsequent releases?
